google/cloud/spanner_v1/snapshot.py

```diff
@@ -447,34 +452,38 @@ def execute_sql(
         if self._transaction_id is None:
             # lock is added to handle the inline begin for first rpc
             with self._lock:
-                iterator = _restart_on_unavailable(
-                    restart,
-                    request,
-                    "CloudSpanner.ReadWriteTransaction",
-                    self._session,
-                    trace_attributes,
-                    transaction=self,
+                return self._get_streamed_result_set(
+                    restart, request, trace_attributes, False
                 )
-                self._read_request_count += 1
-                self._execute_sql_count += 1
-
-                if self._multi_use:
-                    return StreamedResultSet(iterator, source=self)
-                else:
-                    return StreamedResultSet(iterator)
         else:
-            iterator = _restart_on_unavailable(
-                restart,
-                request,
-                "CloudSpanner.ReadWriteTransaction",
-                self._session,
-                trace_attributes,
-                transaction=self,
+            return self._get_streamed_result_set(
+                restart, request, trace_attributes, True
             )
 
+    def _get_streamed_result_set(
+        self, restart, request, trace_attributes, transaction_id_set
+    ):
+        iterator = _restart_on_unavailable(
+            restart,
+            request,
+            "CloudSpanner.ReadWriteTransaction",
+            self._session,
+            trace_attributes,
+            transaction=self,
+        )
         self._read_request_count += 1
         self._execute_sql_count += 1
 
+        if self._read_only and not transaction_id_set:
+            peek = next(iterator)
```

Comment:
> It looks to me like line 485 is effectively rewinding the iterator? So this does implement the peek idiom (assuming no exception is thrown between here and there), though I'd agree that it's confusing.

Yes thats what I am doing

> 1. Previously, calling this method would not actually execute the query. Instead, the query would be executed the first time that next(iterator) is called.
> 2. Now the query is executed directly when this method is called. That is more idiomatic and what most users would expect, but it is also a breaking change if this was not the behavior before this change.

Hmmm, thanks for catching this. Yes I think now we would be executing the query twice on Spanner. 

> Reading the implementation of _restart_on_unavailable, though: Do I read correctly that it's implemented by buffering the full resultset in memory? I'm curious if you know why we do that rather than yielding results as they arrive and, in case of ServiceUnavailable, just replay to the point where we left off and keep streaming? The semantics would be slightly different, so there might be a good reason for this choice.

I went through the function couple of times and its a bit confusing to me. IIUC, its a generator function which is yielding the result on [line 134](https://github.com/googleapis/python-spanner/blob/main/google/cloud/spanner_v1/snapshot.py#L134) but I am not sure if it being generator helps (other than when we get resume_token in the response and need to make subsequent call to Spanner but I see that there is some issue here) as on first call it will make a query to Spanner and get list of PartialResultSet in the response, which it needs to store in memory and the next call will just return it from memory. 

In case of ServiceUnavailable, i can see that we are maintaining resume_token and making call based on that.

Probably I need to understand this function better and how this generator is used in [StreamedResultSet](https://github.com/googleapis/python-spanner/blob/main/google/cloud/spanner_v1/streamed.py) and the how StreamedResultSet is used in [PeekIterator](https://github.com/googleapis/python-spanner/blob/main/google/cloud/spanner_dbapi/utils.py) later when fetch calls are being made on Cursor

Reply:
I have removed this logic from here and moved it to Cursor class where we were already peeking at the iterator